### PR TITLE
Enable lifecycle method testing in enzyme

### LINKED ISSUE #4794: Lifecycle method testing is disabled in the enzyme config
https://github.com/zooniverse/Panoptes-Front-End/blob/fb4b7f8ee9e5f9348b59333e763ccaadcb3715c2/test/enzyme-configuration.js#L6

Enzyme 3 supports testing of `componentDidMount` when shallow rendering, but this option is turned off in our enzyme config. Turning it on breaks some component tests. We should fix those tests then enable this option.

### PR DESCRIPTION
**Pull request by eatyourgreens**
Staging branch URL: https://enzyme-lifecycle-tests.pfe-preview.zooniverse.org

Fixes #4794 

Turns on lifecycle method testing and fixes broken tests.

With this PR, `shallow()` in Enzyme 3 runs `componentDidMount`. This broke some tests that ran `componentDidMount()` manually. It also means any spies that spy on method calls during mount must be defined before running `shallow()`, not after. There were also some components where `componentDidMount` broke if called with the default component props.

# Required Manual Testing

- [x] Does the non-logged in home page render correctly?
- [x] Does the logged in home page render correctly?
- [x] Does the projects page render correctly?
- [x] Can you load project home pages?
- [x] Can you load the classification page?
- [x] Can you submit a classification?
- [x] Does talk load correctly?
- [x] Can you post a talk comment?

# Review Checklist

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [x] Can you `rm -rf node_modules/ && npm install` and app works as expected?
- [ ] If the component is in coffeescript, is it converted to ES6? Is it free of eslint errors? Is the conversion its own commit?
- [x] Are the tests passing locally and on Travis?

# Optional

- [ ] Have you replaced any `ChangeListener` or `PromiseRenderer` components with code that updates component state?
- [ ] If changes are made to the classifier, does the dev classifier still work?
- [ ] Have you [resized and compressed](https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/image-optimization) any image you've added?
- [ ] Have you added in [flow type annotations](https://flowtype.org/docs/type-annotations.html)?
- [ ] Have you followed the [Springer guidelines for commit messages](https://github.com/springernature/frontend-playbook/blob/master/git/git.md#commit-messages)?

### CODE REVIEW

File: app/pages/admin/user-settings.spec.js [hunk 1]
```diff
@@ -1,12 +1,14 @@
...
 import { shallow } from 'enzyme';
 import UserSettings from './user-settings';
 
 describe('UserSettings', function () {
   let wrapper;
-  let user = {id: 1, login: "tester"};
-  let editUser = {id: 2, login: 'volunteer'};
+  let user = {id: '1', login: "tester"};
+  let editUser = {id: '2', login: 'volunteer'};
+  sinon.stub(UserSettings.prototype, 'getUser').callsFake(() => null);
```
Comment (srallen) [inline app/pages/admin/user-settings.spec.js]: Could you put this in a `before` block and add an `after` that restores the method to the original? There could be unintended side effects if the class method isn't restored in other unrelated tests

File: app/pages/notifications/notification-section.spec.js [hunk 2]
```diff
@@ -27,6 +28,10 @@ const notifications = [
 describe('Notification Section', function() {
   let wrapper;
 
+  before(function () {
```
Comment (srallen) [inline app/pages/notifications/notification-section.spec.js]: Could you add an after block that restores the class method to its original unwrapped method?

File: app/talk/active-users.spec.js [hunk 3]
```diff
@@ -11,13 +11,12 @@ const users = [
 ];
 
 describe('ActiveUsers', function () {
+  const getActiveIdsStub = sinon.stub(ActiveUsers.prototype, 'getActiveUserIds').callsFake(() => Promise.resolve(activeIds));
+  const fetchUsersSpy = sinon.spy(ActiveUsers.prototype, 'fetchUncachedUsers');
+  const pageCountSpy = sinon.spy(ActiveUsers.prototype, 'pageCount');
+  const boundedPageSpy = sinon.spy(ActiveUsers.prototype, 'boundedPage');
+  const userIdSpy = sinon.spy(ActiveUsers.prototype, 'userIdsOnPage');
```
Comment (srallen) [inline app/talk/active-users.spec.js]: Same comment about putting these in a `before` block and adding an `after` one to restore these.

File: app/pages/admin/project-status.spec.js [hunk 4]
```diff
@@ -40,6 +40,8 @@ describe('ProjectStatus', function () {
   let wrapper;
 
   before(function () {
+    sinon.stub(ProjectStatus.prototype, 'getProject').callsFake(() => Promise.resolve(project));
+    sinon.stub(ProjectStatus.prototype, 'getWorkflows').callsFake(() => Promise.resolve(workflows));
```
Comment (srallen) [inline app/pages/admin/project-status.spec.js]: Same comment about restoring these in the `after` block.

**Comment by eatyourgreens** [comment]
Good call on cleaning up after tests. I've gone through and updated those spies.

Running the tests without a network connection shows that there are still some components where the API client hasn't been mocked, which explains the long pause when the tests finish running.
```
(node:5570) UnhandledPromiseRejectionWarning: Error: getaddrinfo ENOTFOUND panoptes-staging.zooniverse.org panoptes-staging.zooniverse.org:443
  at errnoException (dns.js:50:10)
  at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:92:26)

(node:5570) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 3)
(node:5570) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
(node:5570) UnhandledPromiseRejectionWarning: Error: getaddrinfo ENOTFOUND panoptes-staging.zooniverse.org panoptes-staging.zooniverse.org:443
  at errnoException (dns.js:50:10)
  at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:92:26)

(node:5570) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 4)
```

**Comment by srallen** [comment]
@eatyourgreens I turned off wifi and ran the tests on master. Some, though not all, of those Promise rejections happening on the master branch anyway, so I'm ok with those being addressed in a separate PR since they may be difficult to track down.

**Pull request merged**